Docs to change Container runtime (#30141)
kubernetes/website: content/en/docs/tasks/administer-cluster/migrating-from-dockershim/change-runtime-containerd.md.md

---

> title: "Migrating Container runtime from dockershim to containerd"

Reviewer: Is dockershim a runtime or a CRI? Maybe "Migrate container runtime from Docker to containerd"

> title: "Migrating Container runtime from dockershim to containerd"

Reviewer 2: nit:
```suggestion
title: "Change Your Container Runtime from Docker Engine to containerd"
```

`dockershim` isn't the runtime, it literally is just the shim.

Reviewer 3: maybe `dockerd` ?

Reviewer 2: @afbjorklund is “Docker Engine” not the right term (maybe that only covers commercially-licensed Docker)?

Usually we use the proper name of a component (eg Linux, Windows) and not the name of the binary.

Reviewer 3: I think Docker Engine is the (current) name of the product, possibly also applies if you compile it yourself from sources

In this context, maybe even `docker` (to match `docker://` in the CLI output) ?

Reviewer: I like Docker as well for this

Reviewer 2: I'm fine with Docker too; from context, we're clearly talking _here_ about Docker as a container runtime.
Also it avoids having to ask what you call the kind of runtime you specify in `/etc/docker/daemon.json`, if Docker counts as a container runtime

> title: "Change Your Container Runtime from Docker Engine to containerd"

Reviewer 4: ```suggestion
title: "Changing the container runtime from Docker Engine to containerd"
```

> weight: 10

Reviewer 2: This weight feels too high. I would make sure this comes after the “do I actually need to migrate” task pages.

content_type: task 
---

> 

Reviewer: I'd highly recommend following the structure in https://github.com/kubernetes/kubernetes/issues/104878#issuecomment-990909830 as it stays in line with our task template and provides contextual information before diving into the tasks. So a brief overview, prerequisites, etc.

Author: Yes the outline will remain the same as you have mentioned and as it is mentioned in the above review here https://github.com/kubernetes/website/pull/30141#discussion_r755274140

Reviewer 4: Please add an introduction to this task.

Also - are there any prerequisites for this task? For example, I assume you need to install containerd. 

If there are please include a "Prerequisites" section before the task steps, i.e. 

```suggestion

This task outlines the steps needed to update your container runtime to containerd from Docker. It is applicable for cluster operators running Kubernetes 1.23 or earlier. 

## Prerequisites 

- Install containerd. For more information see, [containerd's installation documentation](https://containerd.io/docs/getting-started/)
```

Reviewer: +1 (and please use `## {{% heading "prerequisites" %}}` for the heading)

> - Drain Node

Reviewer 2: Do readers need to do this just once, or repeat this for each node? We should make that detail clear.

> ```

Reviewer 4: ```suggestion
```shell
```

# replace <node-to-drain> with the name of your node you are draining
kubectl drain <node-to-drain> --ignore-daemonsets
```

> - Stop the Docker daemon

Reviewer 4: ```suggestion
2. Stop the Docker daemon
```

> ```

Reviewer 4: ```suggestion
```shell
```

systemctl stop kubelet
systemctl stop docker
```

> - Install & start containerd 

Reviewer 4: ```suggestion

2.  Install & start containerd 
```

... Are there any instructions on how to do this here?

> ### For Linux systems

Reviewer 2: ```suggestion
## Migrate a Linux node {#migration-linux}
```

> ## Migrate a Linux node {#migration-linux}

Reviewer 4: Also - are there any reasons why Linux nodes need different install instructions? I ask this because most nodes are Linux nodes.

Use the following commands to install Containerd on your system:

> Install and configure prerequisites:

Reviewer 4: ```suggestion
1. Install and configure prerequisites:
```

```shell
cat <<EOF | sudo tee /etc/modules-load.d/containerd.conf
overlay
br_netfilter
EOF

sudo modprobe overlay
sudo modprobe br_netfilter

# Setup required sysctl params, these persist across reboots.
cat <<EOF | sudo tee /etc/sysctl.d/99-kubernetes-cri.conf
net.bridge.bridge-nf-call-iptables  = 1
net.ipv4.ip_forward                 = 1
net.bridge.bridge-nf-call-ip6tables = 1
EOF

# Apply sysctl params without reboot
sudo sysctl --system

> ```

Reviewer 4: Can we break each one of these chunks into their own steps? 

(sorry, I'd do this as a suggestion but it's breaking GitHub's formatting!)

2. SOME DESCRIPTIVE ACTION STATEMENT

```shell
cat <<EOF | sudo tee /etc/modules-load.d/containerd.conf
overlay
br_netfilter
EOF
```

3. SOMETHING ELSE DESCRIPTIVE

```shell
sudo modprobe overlay
sudo modprobe br_netfilter
```

4. Setup required sysctl params which persist across reboots

```shell
cat <<EOF | sudo tee /etc/sysctl.d/99-kubernetes-cri.conf
net.bridge.bridge-nf-call-iptables  = 1
net.ipv4.ip_forward                 = 1
net.bridge.bridge-nf-call-ip6tables = 1
EOF
```

5. Apply sysctl params without reboot

```shell
sudo sysctl --system
```

Install containerd:

1. Install the `containerd.io` package from the official Docker repositories. 
   You can find instructions for installing `containerd` at
   [Starting Containerd](https://containerd.io/docs/getting-started/#starting-containerd).

2. Configure containerd:

   ```shell
   sudo mkdir -p /etc/containerd
   containerd config default | sudo tee /etc/containerd/config.toml
   ```

3. Restart containerd:

   ```shell
   sudo systemctl restart containerd

>    ```

Reviewer 2: I would explicitly link to the next step the reader should follow, where they start the kubelet etc.

Reviewer: +1 - something like "After you install containerd, you must \[configure the kubelet to use containerd as the container runtime\](#use-containerd-as-runtime)."

> ### For windows powershell

Reviewer 2: ```suggestion
## Migrate a Windows node {#migrate-windows-powershell}
```

Start a Powershell session, set `$Version` to the desired version (ex: `$Version=1.4.3`), 
and then run the following commands:

> 1. Download containerd:

Reviewer: I'd recommend linking to containerd's install instructions instead of duplicating them here.

Reviewer 4: This please ^^. I consider this a blocking change and will not merge until it's done! ❤️‍🔥

Reviewer: The containerd docs don't currently have instructions, but both these sections in this PR are duplicates of our [container runtimes instructions on kubernetes.io](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#containerd). Maybe instead of the duplicate content, we can link to that page? The instructions are the same and are tabbed for readability. 

This would mean our instructions on *this* page would go like:

For each node: 

1. Cordon and drain the node 
2. Install containerd (link)
3. Configure the kubelet
4. Uncordon the node
5. Verify Pods are running
6. Describe the node to verify the runtime

WDYT?

   ```powershell
   curl.exe -L https://github.com/containerd/containerd/releases/download/v$Version/containerd-$Version-windows-amd64.tar.gz -o containerd-windows-amd64.tar.gz
   tar.exe xvf .\containerd-windows-amd64.tar.gz
   ```

2. Extract and configure:

   ```powershell
   Copy-Item -Path ".\bin\" -Destination "$Env:ProgramFiles\containerd" -Recurse -Force
   cd $Env:ProgramFiles\containerd\
   .\containerd.exe config default | Out-File config.toml -Encoding ascii

   # Review the configuration. Depending on setup you may want to adjust:
   # - the sandbox_image (Kubernetes pause image)
   # - cni bin_dir and conf_dir locations
   Get-Content config.toml

   # (Optional - but highly recommended) Exclude containerd from Windows Defender Scans
   Add-MpPreference -ExclusionProcess "$Env:ProgramFiles\containerd\containerd.exe"
   ```

3. Start containerd:

   ```powershell
   .\containerd.exe --register-service
   Start-Service containerd
   ```

> - configure kubelet to use containerd

Reviewer 2: ```suggestion
## Configure the kubelet to use containerd as its container runtime {#use-containerd-as-runtime}
```

>    Edit the file `/var/lib/kubelet/kubeadm-flags.env` and add the containerd runtime to the flags. `--container-runtime=remote` and `--container-runtimeendpoint=unix:///run/containerd/containerd.sock"`

Reviewer 2: Does this apply equally to Linux and Windows systems? If not, let's clarify.

Author: It's different for both the environments

Reviewer 5: it's `--container-runtime-endpoint` with a `-` before `endpoint`.

   For users using kubeadm should consider the following:
   
   Kubeadm stores the CRI socket for each host as an annotation in the Node object for that host.
   To change it you must do the following:
   ```
      - Execute `kubectl edit no <NODE-NAME>` on a machine that has the kubeadm `/etc/kubernetes/admin.conf` file.
      
      This will start a text editor where you can edit the Node object.
      

>       To choose a text editor you can set the `KUBE_EDITOR` environment variable.

Reviewer 2: I think you can also set `EDITOR` - have I got that right?

          - Change the value of `kubeadm.alpha.kubernetes.io/cri-socket` from `/var/run/dockershim.sock`to the CRI socket path of your chose (for example `unix:///run/containerd/containerd.sock`).
   
   Note that new CRI socket paths must be prefixed with `unix://` ideally.

>     - Save the changes in the text editor, which will update the   Node object.

Reviewer 2: I wouldn't put this text in a code block (it's not code).

Author: This was due to my editor not showing me proper preview will change this

   ```
      
   

- restart kubelet
`systemctl start kubelet`

- verify pods are running 
Run `kubectl get nodes -o wide` and containerd appears as the runtime for the node we just changed.

> - finally if everything goes well remove docker

Reviewer 2: How would a reader know if things have gone well?

---
title: "Changing the container runtime from Docker Engine to containerd"
weight: 8
content_type: task 
---
This task outlines the steps needed to update your container runtime to containerd from Docker. It is applicable for cluster operators running Kubernetes 1.23 or earlier. 

## Prerequisites 
1. Install containerd. For more information see, [containerd's installation documentation](https://containerd.io/docs/getting-started/)

2. Drain Node 
   ```
   # replace <node-to-drain> with the name of your node you are draining
   kubectl drain <node-to-drain> --ignore-daemonsets
   ```
3. Stop the Docker daemon
   ```
   systemctl stop kubelet
   systemctl stop docker
   ```

> 2.  Start containerd 

Reviewer: I don't think these are prereqs. Outside of containerd, what else do users need to have ready before they proceed?

> ## Migrate a Linux node {#migration-linux}

Reviewer: I'd change this to something like:

\## Install containerd

To begin migrating your runtime, download and install containerd on your system.

\### Migrate a Linux node

*instructions*

\### Migrate a Windows node

*instructions*

Use the following commands to install Containerd on your system:

1. Install and configure prerequisites:

   ```shell
   cat <<EOF | sudo tee /etc/modules-load.d/containerd.conf
   overlay
   br_netfilter
   EOF
   ```
   ```shell
   sudo modprobe overlay
   sudo modprobe br_netfilter
   ```

> 2. Setup required sysctl params, these persist across reboots.
>    ```shell
>       cat <<EOF | sudo tee /etc/sysctl.d/99-kubernetes-cri.conf
>    net.bridge.bridge-nf-call-iptables  = 1
>    net.ipv4.ip_forward                 = 1
>    net.bridge.bridge-nf-call-ip6tables = 1
>    EOF
>    ```

Reviewer: ```suggestion
1. Set up required sysctl parameters:
   \```shell
      cat <<EOF | sudo tee /etc/sysctl.d/99-kubernetes-cri.conf
   net.bridge.bridge-nf-call-iptables  = 1
   net.ipv4.ip_forward                 = 1
   net.bridge.bridge-nf-call-ip6tables = 1
   EOF
  \```
   These parameters persist across reboots.
```

> 3. Apply sysctl params without reboot

Reviewer: ```suggestion
3. Apply sysctl parameters without reboot:
```

   ```shell
   sudo sysctl --system
   ```

> Install containerd:

Reviewer: ```suggestion

```

1. Install the `containerd.io` package from the official Docker repositories. 
   You can find instructions for installing `containerd` at
   [Starting Containerd](https://containerd.io/docs/getting-started/#starting-containerd).

> 2. Configure containerd:

Reviewer: ```suggestion
1. Configure containerd:
```

   ```shell
   sudo mkdir -p /etc/containerd
   containerd config default | sudo tee /etc/containerd/config.toml
   ```

> 3. Restart containerd:

Reviewer: ```suggestion
1. Restart containerd:
```

   ```shell
   sudo systemctl restart containerd
   ```

> ## Migrate a Windows node {#migrate-windows-powershell}

Reviewer: See comment above about changing the organization of these headings

---
title: "Changing the container runtime from Docker Engine to containerd"
weight: 8
content_type: task 
---
This task outlines the steps needed to update your container runtime to containerd from Docker. It is applicable for cluster operators running Kubernetes 1.23 or earlier. 

> ## Prerequisites 

Reviewer: ```suggestion
## {{% heading "prerequisites" %}}
```
